qualtran/_infra/gate_with_registers.py

```diff
@@ -311,23 +310,40 @@ def on_registers(
     ) -> cirq.Operation:
         return self.on(*merge_qubits(self.signature, **qubit_regs))
 
+    def __pow__(self, power: int) -> 'GateWithRegisters':
+        if power == -1:
+            return self.adjoint()
+        if power == 1:
+            return self
+        raise NotImplementedError("pow is not implemented.")
+
     # pylint: disable=arguments-renamed
     def controlled(
         self,
-        num_controls: Optional[int] = None,
+        num_controls: Union[Optional[int], 'CtrlSpec'] = None,
         control_values=None,
         control_qid_shape: Optional[Tuple[int, ...]] = None,
-    ) -> 'cirq.Gate':
-        from qualtran.cirq_interop import BloqAsCirqGate
-
-        controlled_gate = cirq.ControlledGate(
-            self,
-            num_controls=num_controls,
-            control_values=control_values,
-            control_qid_shape=control_qid_shape,
-        )
-        ctrl_spec = CtrlSpec.from_cirq_cv(controlled_gate.control_values)
-        return BloqAsCirqGate(Controlled(self, ctrl_spec))
+    ) -> 'GateWithRegisters':
```

Comment:
Would it be useful to provide overloaded type hints?
```py
    # pylint: disable=arguments-renamed
    @overload
    def controlled(
        self,
        num_controls: Optional[int] = None,
        control_values=None,
        control_qid_shape: Optional[Tuple[int, ...]] = None,
    ) -> 'GateWithRegisters':
        ...

    # pylint: disable=signature-differs
    @overload
    def controlled(self, ctrl_spec: 'CtrlSpec') -> 'GateWithRegisters':
        ...

    # pylint: disable=arguments-renamed
    def controlled(self, num_controls=None, control_values=None, control_qid_shape=None):
```

Right now it's difficult to understand the usage without looking at the function body.

Reply:
yeah, I think this needs a big docstring that describes the two modes of operation

Reply:
Added a detailed docstring and the overloads. Also added tests to verify the different cases raise errors.